### ISSUE TITLE
JsonSerializer.Deserialize throws ArgumentOutOfRangeException or JsonException reading json file greater than 16KB on .net7.0 but not on .net6.0

### ISSUE THREAD

**Issue by frabe1579** [dotnet/runtime#78692]
### Description

Environment is .net7, version 7.0.0
Reading a json file of 16+KB throws an exception.
The original file is 300+KB, the exception is of type ArgumentOutOfRangeException, and when i reduce it's size to be just a bit more than 16KB, the exception is of type JsonException.
If the file is less than 16KB, no exception is thrown.
The attached [json.zip](https://github.com/dotnet/runtime/files/10066068/json.zip) has three files:
- italy_cap0.json is less than 16KB, is read without problems;
- italy_cap1.json is a bit more than 16KB, throws JsonException;
- italy_cap2.json is 380+KB, throws ArgumentOutOfRangeException;

If the project is downgraded to .net6 version 6.0.11 no exception is thrown in any case.

This is the example code for a Console program:
```
using System.Runtime.InteropServices;
using System.Text.Json;

Console.WriteLine(RuntimeInformation.FrameworkDescription);

var stm2 = File.OpenRead(@"italy_cap2.json");
var caps = JsonSerializer.Deserialize<List<CapInfo>>(stm2, new JsonSerializerOptions { PropertyNameCaseInsensitive = true })!;

Console.WriteLine("File read successfully!");

public class CapInfo
{
  public CapInfo(string istat, string cap)
  {
    Istat = istat;
    Cap = cap;
  }

  public string Istat { get; }
  public string Cap { get; }
}
```

If I read the content of the file in a string before deserializing (using File.ReadAllText), and then passing the string to the deserializer, no exception is thrown.

For italy_cap1.json (16+KB) the stack trace is the following:
```
Unhandled exception. System.Text.Json.JsonException: 's' is an invalid start of a property name. Expected a '"'. Path: $[334] | LineNumber: 1338 | BytePositionInLine: 3.
 ---> System.Text.Json.JsonReaderException: 's' is an invalid start of a property name. Expected a '"'. LineNumber: 1338 | BytePositionInLine: 3.
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
   at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
   at System.Text.Json.Utf8JsonReader.Read()
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   --- End of inner exception stack trace ---
   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, JsonReaderException ex)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ContinueDeserialize[TValue](ReadBufferState& bufferState, JsonReaderState& jsonReaderState, ReadStack& readStack, JsonTypeInfo jsonTypeInfo)
   at System.Text.Json.JsonSerializer.ReadFromStream[TValue](Stream utf8Json, JsonTypeInfo jsonTypeInfo)
```
For italy_cap2.json (300+KB) the stack trace is the following:
```
Unhandled exception. System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
   at System.Text.Json.JsonSerializer.ContinueDeserialize[TValue](ReadBufferState& bufferState, JsonReaderState& jsonReaderState, ReadStack& readStack, JsonTypeInfo jsonTypeInfo)
   at System.Text.Json.JsonSerializer.ReadFromStream[TValue](Stream utf8Json, JsonTypeInfo jsonTypeInfo)
```

### Reproduction Steps

- Create a console project for .net 7 as this:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <None Update="italy_cap0.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="italy_cap1.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="italy_cap2.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
  </ItemGroup>

</Project>
```
- Copy the json files of the attached zip to the same directory of the project.
- Set this as Program.cs:
```
using System.Runtime.InteropServices;
using System.Text.Json;

Console.WriteLine(RuntimeInformation.FrameworkDescription);

var stm2 = File.OpenRead(@"italy_cap2.json");
var caps = JsonSerializer.Deserialize<List<CapInfo>>(stm2, new JsonSerializerOptions { PropertyNameCaseInsensitive = true })!;

Console.WriteLine("File read successfully!");

public class CapInfo
{
  public CapInfo(string istat, string cap)
  {
    Istat = istat;
    Cap = cap;
  }

  public string Istat { get; }
  public string Cap { get; }
}
```

### Expected behavior

No exception is thrown, json is read as a list of objects.

### Actual behavior

An exception is thrown, apparently depending of the file size.

### Regression?

Yes, works in .net6 version 6.0.11

### Known Workarounds

Read the file as a string before deserializing, and then call the appropriate overload, instead of passing the stream to JsonSerializer.Deserialize<>(...)

### Configuration

.net version 7.0.0
Windows 10 Pro x64
Version	21H2
Build 19044.2251
Windows Feature Experience Pack 120.2212.4180.0

### Other information

The problem seems to be, but is a conjecture, unaligned buffering of input stream.

**Comment by eiriktsarpalis** [comment]
Duplicate of #78221. The fix has been backported to .NET 7 which should be released in December. Nightly builds for .NET 7.0.2 are also [available to download](https://github.com/dotnet/installer#table) right now.

**Comment by Nuklon** [comment]
@eiriktsarpalis I cannot find this in the changelog, or under 7.0.2 approved issues, is this fix part of 7.0.1?

**Comment by eiriktsarpalis** [comment]
Yes, looks like the fix was released in 7.0.1. Apologies for the confusion.
